soc/riscv/riscv-privilege/virt/soc.c

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright (c) 2021 Katsuhiro Suzuki
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief QEMU RISC-V virt machine hardware depended interface
+ */
+
+#include <kernel.h>
+#include <arch/cpu.h>
+#include <sys/util.h>
+
+/*
+ * Exit QEMU and tell error number.
+ *   Higher 16bits: indicates error number.
+ *   Lower 16bits : set FINISHER_FAIL
+ */
```

Comment:
Do you know if there is a specification for this interface? I did a quick search and could only find it referenced in the QEMU implementation, but unfortuantely no document.
If there is a document describing the interface, then I could be referenced in a comment here. Something like: "This implements the _SiFive test finisher_, described in document X, section Y.Z".

Reply:
Yes, good points. Unfortunately, no documentation...
Here is QEMU patch for adding this feature, they added no documents with this implementation.
https://patchwork.kernel.org/project/qemu-devel/patch/1519998711-73430-18-git-send-email-mjc@sifive.com/